### DISCUSSION TITLE
How to stop Corfu from stealing the DEL key?

### DISCUSSION THREAD

**Discussion by egr95** [minad/corfu#454] [Q&A] [answered]
Suppose I use `TAB` (possibly repeatedly) to select one of the autocompletion candidates. If I then hit `DEL`, it removes the completion altogether, instead of simply deleting the last character in the completion. How could I stop Corfu from stealing `DEL`, so that it just accepts the completion and deletes the last character? If I want to cancel the completion I can already press `C-g` instead of `DEL`.

Essentially what I would like is for `TAB` to select the autocompletion candidates (with cycling and `S-TAB` for reversing), `C-g` to cancel autocompletion, and any other operation to just "pass through" corfu, so in other words any other keypress should simply accept the completion and do whatever it would have done anyway. Is this something I can configure more easily than manually disabling all the keys that corfu takes over?

I have tried all of the following but to no avail:

```
  (keymap-unset corfu-map "<remap> <backward-delete-char-untabify>")
  (keymap-unset corfu-map "<remap> <backward-delete-char>")
  (keymap-set corfu-map "DEL" 'nil)
  (keymap-set corfu-map "<deletechar>" 'nil)
  (define-key corfu-map [deletechar] nil)
  (define-key corfu-map [remap deletechar] nil)
  (define-key corfu-map [?\d] nil)
  (define-key corfu-map [remap ?\d] nil)
```

Thanks!

**Comment by minad** [comment]
Hi! I am not sure if I understand, since Corfu doesn't steal the DEL key. See the bindings in `corfu-map`. Maybe the interaction with `corfu-preview-current` is leading to confusion here? Try to set `corfu-preview-current` to nil. Does the behavior make more sense then?

**Comment by egr95** [comment]
I see. This is indeed to do with `corfu-preview-current`. I thought the preview of the current selection was in fact being written into the buffer, but it seems that this does not happen until the selection is somehow "accepted". Basically I would like for this "acceptance" to happen automatically upon any subsequent keypress. This already happens with most keys, for example, if I `TAB` into the autocompletion choices and then type any letter of the alphabet, then corfu automatically accepts what I had selected, and the subsequent action applies "post-selection". Is there a natural way to configure this in Corfu? 

Bottom line, I would like `DEL` (and essentially all keys other than `TAB`, `S-TAB`, `M-SPC` and `C-g`) to automatically accept the previewed selection.

**Reply by minad** [reply]
> Bottom line, I would like DEL (and essentially all keys other than TAB, S-TAB, M-SPC and C-g) to automatically accept the previewed selection.

Yes, this is what already happens with `corfu-preview-current=insert`, which is the Corfu default.

**Reply by egr95** [reply]
That's odd - this is not the behavior I am seeing in Doom emacs. For me, DEL actually cancels the selection rather than deleting its last character. Let me dig up the doom module to see if this might be as a result of the Doom configuration of Corfu.

**Reply by minad** [reply]
Maybe @lemonbreezes can help?

**Reply by LemonBreezes** [reply, marked as answer]
To disable Doom Emacs's Corfu backspace behavior, evaluate:
```elisp
(after! corfu
  (map! :map corfu-map
        [backspace] nil
        "DEL" nil))
```

**Reply by egr95** [reply]
Worked a treat, thank you @LemonBreezes !
